doc/tsm/tsi.md

```diff
@@ -35,19 +33,49 @@ These measurement blocks may duplicate measurement blocks created earlier by oth
 
 As TSI files grow and measurements are duplicated, query time merging becomes more costly so a larger compaction will be necessary to consolidate all series, measurement, and tag value blocks into one.
 
+When an index file or WAL file is compacted it is locked and the new index and WAL will begin on a new file. This is to allow snapshotting to take hard links on the files.
+
+#### Two Compaction Types
+
+There are two separate types of compaction that can occur:
+
+1. Fast compaction -- compacts a set of WAL entries into series, measurement, and tag value blocks. These blocks are layered on top of previously compacted block sets.
+
+2. Full compaction -- compacts sets of block sets into a single block set. This takes the multiple layers of series, measurement, and tag value blocks and merges them into one. It also removes any entries which have been tombstoned.
+
 
 ## Detailed Block Information
 
 ### Series Blocks
 
-Series blocks contain a `uint32` block ID followed by a list of `uint32` local series IDs. Combining the block id and local series ID provides a globally addressable `uint64` series ID: `(blockID,seriesID)`
+Series blocks contain a `uint32` block ID followed by a list of `uint32` local series IDs. Combining the block id and local series ID provides a globally addressable `uint64` series ID: `(blockID,seriesID)`. Block IDs are assigned in sequential order and are only unique within an index file.
+
+For example, if the block ID was 1 and the local series ID was 5 then the global ID is (1,5) and is represented as `1 << 32 | 5`.
 
 The layout on disk for the block would be:
 
 ```
 BLOCKID<uint32>
 KEYN<uint32>
 (KEYPOS<uint64>)*
+(FLAG<uint8>,KEYSZ<uint16>,KEY<string>)*
+```
+
+With a global ID, the key can be looked up by moving to the appropriate block and then moving to the `KEYPOS` based on the local series ID and then finally jumping to the key itself using the key position.
+
+The `FLAG` contains boolean flags for the series. The first bit is a tombstone flag to mark the series as deleted. Other bits are reserved for future use.
```

Comment:
The blocks for an index are immutable. Fast compactions will create new blocks which will be searched first before going to older blocks. For example, if you had two index files:

```
┌───────────────────────────────────────────┐          ┌───────────────────────────────────────────┐
│                   0.tsi                   │          │                   1.tsi                   │
│                                           │          │                                           │
│    ┌─────────────────────────────────┐    │          │    ┌─────────────────────────────────┐    │
│ ┌─▶│          Series Block           │    │          │ ┌─▶│          Series Block           │    │
│ │  ├─────────────────────────────────┤    │          │ │  ├─────────────────────────────────┤    │
│ ├──│   Tag Values Block (cpu.pid)    │◀─┐ │          │ └──│  Tag Values Block (mem.region)  │◀─┐ │
│ │  ├─────────────────────────────────┤  │ │          │    ├─────────────────────────────────┤  │ │
│ ├──│  Tag Values Block (cpu.region)  │◀─┤ │          │    │     Measurement Block (mem)     │──┘ │
│ │  ├─────────────────────────────────┤  │ │          │    └─────────────────────────────────┘    │
│ │  │     Measurement Block (cpu)     │──┘ │          │                                           │
│ │  ├─────────────────────────────────┤    │          └───────────────────────────────────────────┘
│ │  │  Tag Values Block (mem.region)  │◀─┐ │                                                       
│ │  ├─────────────────────────────────┤  │ │                                                       
│ └──│     Measurement Block (mem)     │──┘ │                                                       
│    └─────────────────────────────────┘    │                                                       
│                                           │                                                       
└───────────────────────────────────────────┘                                                       
```

The `1.tsi` index would need to be searched first and then the `0.tsi` index would be searched. Once those files are created they don't change. A series created in `0.tsi` can be flagged as tombstoned in `1.tsi`.

Once a full compaction occurs then the tombstoned series can be deleted entirely and a `2.tsi` will be written. I'm not married to any naming scheme. Maybe the WAL index should be used in the name?